core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/SimplifyFilterPredicate.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.metadata.Metadata;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.tree.Cast;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.IfExpression;
+import io.trino.sql.tree.IsNullPredicate;
+import io.trino.sql.tree.LogicalBinaryExpression;
+import io.trino.sql.tree.NotExpression;
+import io.trino.sql.tree.NullIfExpression;
+import io.trino.sql.tree.NullLiteral;
+import io.trino.sql.tree.SearchedCaseExpression;
+import io.trino.sql.tree.SimpleCaseExpression;
+import io.trino.sql.tree.WhenClause;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.sql.ExpressionUtils.combineConjuncts;
+import static io.trino.sql.ExpressionUtils.extractConjuncts;
+import static io.trino.sql.planner.plan.Patterns.filter;
+import static io.trino.sql.tree.BooleanLiteral.FALSE_LITERAL;
+import static io.trino.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static io.trino.sql.tree.LogicalBinaryExpression.Operator.AND;
+import static io.trino.sql.tree.LogicalBinaryExpression.Operator.OR;
+
+/**
+ * Simplify conditional expressions in filter predicate.
```

Comment:
Why is this specific to Filter node? It should apply to any expressions in any context ,no?

It might be better to fold this into `SimplifyExpressions`, instead.

Reply:
I considered folding it into `SimplifyExpressions`, but found that `SimplifyExpressions` is designed for general expression optimization (i.e. for equivalent transformations).

The expression transformations in this rule don't apply to other contexts than filter. They are based on the observation that filter only cares if the expression is `true` or not `true`. As a consequence, it is safe to lose the distinction between `false` and `null`. The resulting expression might not be equivalent to the original expression.
E.g. `IF(a, true, null)` is transformed into `a`. The result is different for `a == false`, but the filtering effect is the same.